cmake: Get rid of undocumented `BITCOIN_GENBUILD_NO_GIT` environment variable

hebasto opened the pull request:
In general, the Bitcoin Core build system attempts to fetch commit or tag details from git. This is handled by the [`cmake/script/GenerateBuildInfo.cmake`](https://github.com/bitcoin/bitcoin/blob/master/cmake/script/GenerateBuildInfo.cmake) script, which generates the [`src/bitcoin-build-info.h`](https://github.com/bitcoin/bitcoin/blob/65dcbec75661d1652beb927f03b1feab4fab932e/src/clientversion.cpp#L26-L31) header within the build tree.

However, there are cases where the retrieved details may be incorrect—for example, when building from a source tarball or as a subproject within a git-aware project.

In the Autotools-based build system, the `BITCOIN_GENBUILD_NO_GIT` environment variable was [introduced](https://github.com/bitcoin/bitcoin/pull/7522) in [v0.20.0](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.20.0.md) to address such scenarios:
> The process for generating the source code release ("tarball") has changed in an effort to make it more complete, however, there are a few regressions in this release:
> - Instead of running `make` simply, you should instead run `BITCOIN_GENBUILD_NO_GIT=1 make`.

This PR automagically handles both of the aforementioned cases and removes the need for `BITCOIN_GENBUILD_NO_GIT`.

The user is still able to configure the build with [`-DCMAKE_DISABLE_FIND_PACKAGE_Git=ON`](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html) to disable git execution manually, for [reasons](https://github.com/bitcoin/bitcoin/pull/32220#issuecomment-2780115034) we don't know in advance.

Closes https://github.com/bitcoin/bitcoin/issues/31999.

luke-jr commented:
Concept NACK to removing the ability for the user to disable `git` execution manually (perhaps for reasons we don't know in advance).

luke-jr commented on cmake/script/GenerateBuildInfo.cmake:
```diff
@@ -28,9 +28,12 @@ else()
-set(GIT_COMMIT)
-if(NOT "$ENV{BITCOIN_GENBUILD_NO_GIT}" STREQUAL "1")
+set(IS_SOURCE_TARBALL FALSE)
+# git will put "set(IS_SOURCE_TARBALL TRUE)" on the next line inside archives. $Format:%nset(IS_SOURCE_TARBALL TRUE)$
```
Maybe we should use `$Format:%(describe)%$` to actually include build info?

luke-jr replied:
This way we can put the *correct* build info in, rather than just failing to get any build info.

hebasto commented:
> Concept NACK to removing the ability for the user to disable `git` execution manually (perhaps for reasons we don't know in advance).

The user is still able to configure the build with `-DCMAKE_DISABLE_FIND_PACKAGE_Git=ON`.

hebasto commented:
> > The user is still able to configure the build with `-DCMAKE_DISABLE_FIND_PACKAGE_Git=ON`.
> 
> Could mention in the pull description? Also, it looks like the tail needs editing? Ref:
> 
> > This PR handles with both mentioned cases automagically and gets rid of `BITCOIN_GENBUILD_NO_GIT`.
> > This PR automagically handles both of the aforementioned cases and removes the need for `BITCOIN_GENBUILD_NO_GIT`.

Thanks! Updated.

whitslack commented:
> The user is still able to configure the build with [`-DCMAKE_DISABLE_FIND_PACKAGE_Git=ON`](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html) to disable git execution manually, for [reasons](https://github.com/bitcoin/bitcoin/pull/32220#issuecomment-2780115034) we don't know in advance.

But that won't work, will it? The `find_package(Git)` call is in `cmake/script/GenerateBuildInfo.cmake`, which is executed recursively in an independent CMake process by the `generate_build_info` custom target of `src/CMakeLists.txt`. (Did we really learn nothing from "Recursive Make Considered Harmful"? :man_facepalming:) The only reason `BITCOIN_GENBUILD_NO_GIT` was effective in the sub-CMake process is because the environment is inherited by child processes. CMake variable definitions are not inherited.

hebasto commented:
@whitslack 

> > The user is still able to configure the build with [`-DCMAKE_DISABLE_FIND_PACKAGE_Git=ON`](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html) to disable git execution manually, for [reasons](https://github.com/bitcoin/bitcoin/pull/32220#issuecomment-2780115034) we don't know in advance.
> 
> But that won't work, will it? The `find_package(Git)` call is in `cmake/script/GenerateBuildInfo.cmake`, which is executed recursively in an independent CMake process by the `generate_build_info` custom target of `src/CMakeLists.txt`. (Did we really learn nothing from "Recursive Make Considered Harmful"? 🤦‍♂️) The only reason `BITCOIN_GENBUILD_NO_GIT` was effective in the sub-CMake process is because the environment is inherited by child processes. CMake variable definitions are not inherited.

Thank you for your feedback!

Reworked.

theuni commented:
Concept ACK and quick review ACK. Nice to have git archive setting this.

And disabling git from CMake seems like a perfectly reasonable opt-out to me.

fanquake commented:
Concept ACK

maflcko commented on src/CMakeLists.txt:
```diff
@@ -37,9 +37,14 @@ if(NOT CMAKE_ARCHIVE_OUTPUT_DIRECTORY)
 endif()
 
+set(IS_SOURCE_TARBALL FALSE)
+# git will put "set(IS_SOURCE_TARBALL TRUE)" on the next line inside archives. $Format:%nset(IS_SOURCE_TARBALL TRUE)$
```
style nit: maybe rebase on top of faf99ae3796 and fix it here as well

hebasto replied:
Thanks. [Done](https://github.com/bitcoin/bitcoin/pull/32220#issuecomment-4451230329).

hebasto committed:
cmake, refactor: Move `find_package(Git)` to `src/CMakeLists.txt`

This change allows the use of the `CMAKE_DISABLE_FIND_PACKAGE_Git`
variable to explicitly disable git usage.

hebasto committed:
cmake: Remove unnecessary `BITCOIN_GENBUILD_NO_GIT` environment variable

This functionality is now covered by the `CMAKE_DISABLE_FIND_PACKAGE_Git`
variable.

Note for reviewers: consider reviewing with `git diff -w`.

hebasto committed:
cmake: Skip using git when building from source tarball or as subproject

hebasto committed:
doc: Add release notes for #32220

hebasto commented:
> style nit: maybe rebase on top of [faf99ae](https://github.com/bitcoin/bitcoin/commit/faf99ae379636d6ef8316ffb2efaa61896343de9) and fix it here as well

Rebased.

Release notes have been added.

maflcko commented:
review ACK 3142e5f8cf41e3e7fe3eee963cc696765443368e 🥄

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK 3142e5f8cf41e3e7fe3eee963cc696765443368e 🥄
YVYiC4ppg92PxmvnfWlfX0an4AOKrAqcocgGA8nHbQLBYmagLgSVqAmq4XrNKJ5DVVLby59EFKynt9w0J0SYCQ==
```

</details>

fanquake commented:
ACK 3142e5f8cf41e3e7fe3eee963cc696765443368e - tested a few different build scenarios (from Guix the tarball), they all seem to work as intended.
